heudiconv/bids.py

```diff
@@ -496,3 +520,507 @@ def convert_sid_bids(subject_id):
+    criterion : str in ['First', 'Closest']
+        matching_parameters that will be used to decide which of the matching
+        fmaps to use
+    """
+
+    if not isinstance(matching_parameters, list):
+        assert isinstance(matching_parameters, str), "matching_parameters must be a str or a list, got %s" % matching_parameters
+        matching_parameters = [matching_parameters]
+    for param in matching_parameters:
+        if param not in AllowedFmapParameterMatching:
+            raise ValueError(
+                "Fmap matching_parameter %s not allowed." % param
+            )
+    if criterion not in AllowedCriteriaForFmapAssignment:
+        raise ValueError(
+            "Fmap assignment criterion '%s' not allowed." % criterion
+        )
+
+    lgr.info('Adding "IntendedFor" to the fieldmaps in %s.', path_to_bids_session)
+
+    # Resolve path (eliminate '..')
+    path_to_bids_session = op.abspath(path_to_bids_session)
+
+    # Get the subject folder (if "path_to_bids_session" includes the session,
+    # remove it). "IntendedFor" paths will be relative to it.
+    if op.basename(path_to_bids_session).startswith('ses-'):
+        subj_folder = op.dirname(path_to_bids_session)
+    else:
+        subj_folder = path_to_bids_session
+
+    fmap_dir = op.join(path_to_bids_session, 'fmap')
+    if not op.exists(fmap_dir):
+        lgr.warning('We cannot add the IntendedFor field: no fmap/ in %s', path_to_bids_session)
+        return
+
+    compatible_fmaps = find_compatible_fmaps_for_session(
+        path_to_bids_session,
+        matching_parameters=matching_parameters
+    )
+    selected_fmaps = {}
+    for json_file, fmap_groups in compatible_fmaps.items():
+        if not op.dirname(json_file).endswith('fmap'):
+            selected_fmaps[json_file] = select_fmap_from_compatible_groups(
+                json_file,
+                fmap_groups,
+                criterion=criterion
+            )
+
+    # Loop through all the unique fmap_groups in compatible_fmaps:
+    unique_fmap_groups = {}
+    for cf in compatible_fmaps.values():
+        for (key, values) in cf.items():
+            if key not in unique_fmap_groups:
+                unique_fmap_groups[key] = values
+
+    for fmap_group in unique_fmap_groups:
+        intended_for = []
+        for json_file, selected_fmap_group in selected_fmaps.items():
+            if selected_fmap_group and (fmap_group in selected_fmap_group):
+                intended_for.append(
+                    op.relpath(remove_suffix(json_file, '.json') + '.nii.gz', start=subj_folder)
+                )
+        if intended_for:
+            intended_for = sorted([str(f) for f in intended_for])
+            # Add this intended_for to all fmap files in the fmap_group:
+            for fm_json in unique_fmap_groups[fmap_group]:
+                update_json(fm_json, {"IntendedFor": intended_for}, pretty=True)
+
+
+class BIDSFile(object):
```

Comment:
woohoo -- here it is also!  ideally #544 should be finished/merged first (so would be driven by a specific copy of the schema etc), but doesn't need to be.  If this PR is merged first, that one should be rebased/code moved/RFed and would immediately take advantage of this use-case use